File: Greedy/KnapSack.c
#include <stdio.h>

void display(int array[10] , int n);
int knapSack(int W , int weight[10] , int value[10] , int n);

int main(void){
    int W , n ; 
    int weight[10] , value[10];
    printf("Capacity:");
    scanf("%d" , &W);
    printf("No. of items:");
    scanf("%d" , &n);
    printf("Weight array:");
    for(int i = 0 ; i < n ; i++){
        scanf("%d" , &weight[i]);
    }
    printf("Value array:");
    for(int i = 0 ; i < n ; i++){
        scanf("%d" , &value[i]);
    }
    printf("%d",knapSack(W , weight , value , n));
    return 0;
    }

int knapSack(int W , int weight[10] , int value[10] , int n){
    int perKgValue[n];
    for(int i= 0 ; i < n ; i++){
        perKgValue[i] = value[i]/weight[i];
    }
    display(perKgValue , n);
    int profit = 0 , max;  
    while(W){
        max = 0; 
        for(int i= 1 ; i<n ; i++){
            if(perKgValue[i] > perKgValue[max]){
                max = i ; 
            }
        }
        if (weight[max] <= W){
            W -= weight[max];
            profit+= value[max];
        }else{
            profit += perKgValue[max]*W;
            W = 0 ; 
        }
            perKgValue[max] = 0 ; 
    }
    return profit;
}

void display(int array[10] , int n){
    for(int i= 0 ; i < n ; i++){
        printf("%d ",array[i]);
    }
    printf("\n");
}

File: Greedy/KnapSack.py
# There is a thief 
# use karna hai chori 
# ghusa dukan mein 
# uske paas hai bora jiska capacity hai W 
# gaya dukaan mein aur uska pata chala ki kaunsa item kitna weight mein hai aur utne pure weight ka us item ka daam kitna hai 
# weight and value 

def knapSack(W:int , weight: list[int] ,  value:list[int])->int:
    # Fractional Knapsack problem 
    # Aisa tareeka jisse hamara profit maximum ho 
    perKgValue = []
    for i in range(0 , len(weight)):
        # for(int i=0 ; i < len(weight) , i++)
        perKgValue.append(value[i]/weight[i])
        #[4.0, 6.0, 6.0, 2.0, 7.0]
    # maxValue - i W - weight[i] Profit + value[i]
    # if weight[i] > W: W kg 



    
    
knapSack(15 , [5 , 4 , 3 , 2 , 7] , [20 , 24 , 18 , 4 , 49])
